tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,769 @@
+	t.BoolVal = make([]bool, len(data))
+	for i, v := range data {
+		t.BoolVal[i] = (v == 1)
+	}
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return t.BoolVal, nil
+}
+
+// GetVal returns the value of the element contained in the specified position
+// in the tensor, Ex: GetVal(1, 2, 3) is equivalent to data[1][2][3] on a
+// multidimensional array.
+//  This method returns an error if the number of dimensions is incorrect or
+//  are out of range.
+func (t *Tensor) GetVal(d ...int) (val interface{}, err error) {
+	if len(d) != t.NumDims() {
+		return nil, &ErrDimsOutOfTensorRange{
+			tensorDim: t.NumDims(),
+			specDims:  len(d),
+		}
+	}
+
+	pos := 0
+	if t.dimWeights != nil {
+		for i, w := range t.dimWeights {
+			pos += d[i] * w
+		}
+	} else {
+		// Calculate the cumulative weight for each dimension, the
+		// weight is the number of elements before the first of the
+		// elements on this dimension
+		t.dimWeights = make([]int, len(d))
+		pos = d[len(d)-1]
+		if pos >= t.Dim(len(d)-1) {
+			return nil, &ErrIndexOutOfRange{
+				dim:       len(d) - 1,
+				index:     pos,
+				dimsRange: t.Dim(len(d) - 1),
+			}
+		}
+		t.dimWeights[len(d)-1] = 1
+
+		lastWeight := 0
+		for i := len(d) - 2; i >= 0; i-- {
+			lastWeight += t.Dim(i + 1)
+			t.dimWeights[i] = lastWeight
+			pos += d[i] * lastWeight
+
+			if d[i] >= t.Dim(i) {
+				return nil, &ErrIndexOutOfRange{
+					dim:       i,
+					index:     pos,
+					dimsRange: t.Dim(i),
+				}
+			}
+		}
+	}
+
+	return t.getValOnPos(pos)
+}
+
+// getValOnPos returns the value of one of the elements of the Tensor on the
+// specified position
+func (t *Tensor) getValOnPos(pos int) (val interface{}, err error) {
+	switch t.DataType() {
+	case DTFloat:
+		vals, _ := t.Float32()
+		return vals[pos], nil
+	case DTDouble:
+		vals, _ := t.Float64()
+		return vals[pos], nil
+	case DTInt8, DTInt16, DTInt32, DTUint8:
+		vals, _ := t.Int32()
+		return vals[pos], nil
+	case DTInt64:
+		vals, _ := t.Int64()
+		return vals[pos], nil
+	case DTBool:
+		vals, _ := t.Bool()
+		return vals[pos], nil
+	case DTString:
+		vals, _ := t.Str()
+		return vals[pos], nil
+	}
+
+	return nil, &ErrTensorTypeNotSupported{
+		tensotType: t.DataType(),
+	}
+}
+
+// setCMemAsAlreadyRelease indicates that the C allocated memory was already
+// released from C.
+func (t *Tensor) setCMemAsAlreadyRelease() {
+	t.memReleased = true
+}
+
+// FreeAllocMem releases the C allocated memory for this tensor.
+func (t *Tensor) FreeAllocMem() {
+	// We can't clean the tensor here in case it had been  used as an
+	// input parameter, because in tensorflow/core/client/tensor_c_api.cc the
+	// function TF_Run_Helper cleans the input tensors after every
+	// execution. This can cause a double free or corruption error in C++
+	// since there is no way to determine if a tensor had been previously
+	// cleaned.
+	if !t.memReleased {
+		TF_DeleteTensor(t.tensor)
+	}
+}
+
+// ErrInvalidTensorType the data type of the tensor is not compatible
```

Comment:
ErrInvalidTensorType is returned when the data type of the tensor is not compatible...